bayes_kit/iat.py

```diff
@@ -115,43 +113,45 @@ def iat_imse(chain: VectorType) -> FloatType:
     Hall/CRC.
 
     Parameters:
-    chain: Markov chain whose ESS is returned
+    chain: A Markov chain.
 
     Return:
-    estimated effective sample size for the specified Markov chain
+    An estimate of integrated autocorrelation time (IAT) for the specified chain.
 
     Throws:
-    ValueError: if there are fewer than 4 elements in the chain
+    ValueError: If there are fewer than 4 elements in the chain.
     """
     if len(chain) < 4:
         raise ValueError(f"iat requires len(chains) >=4, but {len(chain) = }")
     acor = autocorr(chain)
     n = _end_pos_pairs(acor)
     prev_min = acor[0] + acor[1]
-    # convex minorization to enforce monotonic downward estimates uses slow loop
-    accum = prev_min
+    acor_sum = prev_min
     i = 2
     while i + 1 < n:
+        # enforce monotone downward condition (slow loop)
         prev_min = min(prev_min, acor[i] + acor[i + 1])
-        accum = accum + prev_min
+        acor_sum += prev_min
         i += 2
-    iat = 2 * accum - 1
-    return iat
+    return 2 * acor_sum - 1
 
 
 def iat(chain: VectorType) -> FloatType:
     """
-    Return an estimate of the effective sample size of the specified Markov chain
-    using the default ESS estimator (currently IMSE).  Evaluated by delegating
-    to `iat_imse()`.
+    Return an estimate of the integrated autocorrelation time (IAT)
+    of the specified Markov chain. Evaluated by delegating to the
+    initial monotone sequence estimator, `iat_imse(chain)`.
```

Comment:
Would there be any reason (performance?) someone might want the `ipse` instead? Is it worth exposing the ability to choose that? (I'm guessing not really)

Reply:
I'm OK getting rid of IPSE; even though it's a bit faster than IMSE, it has an overestimation bias.    Do you think I should get rid of it?

I still think we should keep the `ess_imse` and `iat_imse` definitions and delegate `ess` and `iat`.  I know at least for ess we will have further estimators that work cross chain.  The other option is to not provide a default implementation.  I'm always on the fence about choosing default implementations for users, but in the end, I think they're going to want o see `ess()` or `iat()` in most applied code.

Reply:
I don't have a strong feeling about IPSE--it's there, it's implemented, it works; no reason to discard that work, especially if we might want to expose it later. The counterargument is that it's more surface area, and unreachable unused code gets out of date (not that any code in Python is ever truly unreachable).

Unfortunately, I'm kind of clueless whether anybody would in practice actually want to have a choice here.

And if we expose that choice, however we did so, I could imagine a tricky situation where we wind up having some deeply nested calls such that there's an inconsistency between the implementations used in different places--whether because it's hard to make sure the "use-this-implementation" flag gets passed around everywhere `ess` is called, or because users call `ess_ipse` in their own direct code but our version doesn't. Of course, again, dunno if such a scenario is actually likely in practice.

I guess it would come down to three questions:
- Would anybody have a reason to use IPSE?
- Are these two the only implementations worth implementing, or could we imagine adding more later?
- Are these used only/mainly for evaluating sample quality after the fact, or are there cases where a sampling method might want to make use of these in a way that affects its behavior?

Reply:
I talked to @WardBrian and concluded we should leave it in for pedagogical purposes.  

* I think the utility for IPSE is pedagogical and I don't think anyone would use one in an application unless they really needed the extra speed.  The standard way to teach these estimators is to first describe the initial positive sequence method, which was in common use until IPSE was developed, then add a monotonicity condition on top of that.  
* Yes, we'll probably be doing other versions of these eventually, such as a Bayesian estimator that makes a parametric or non-parametric assumption about the shape of the autocorrelation vs. lag curve---it's an active area of research.
* Yes, we will eventually want to use ESS at run time to set a minimum ESS target and run until it's hit (it won't affect the transitions, just when we stop)

Reply:
Sounds good to me. All of this points to the same conclusion--leave it in for pedagogy (and so people could technically call it if they really wanted to), but have a standard pathway that delegates to a specific implementation.